server/container/util/src/main/java/org/apache/james/util/StreamUtils.java

```diff
@@ -54,4 +57,42 @@
     public static <T> Stream<T> flatten(Stream<T>... streams) {
         return flatten(Arrays.stream(streams));
     }
+
+    public static <T> Stream<T> unfold(T seed, Function<T, Optional<T>> generator) {
+        return StreamSupport.stream(new UnfoldSpliterator(seed, generator), false);
+    }
+
+
+    private static class UnfoldSpliterator<T> implements Spliterator<T> {
+
+        private Optional<T> current;
+        private final Function<T, Optional<T>> generator;
+
+        private UnfoldSpliterator(T seed, Function<T, Optional<T>> generator) {
+            this.current = Optional.of(seed);
+            this.generator = generator;
+        }
+
+        @Override
+        public boolean tryAdvance(Consumer<? super T> action) {
+            current.ifPresent(action);
+            current = current.flatMap(generator);
+            return current.isPresent();
+        }
+
+        @Override
+        public Spliterator<T> trySplit() {
+            return null;
```

Comment:
is it dangerous?

Reply:
I don't think so

Reply:
https://docs.oracle.com/javase/8/docs/api/java/util/Spliterator.html#trySplit--

> a Spliterator covering some portion of the elements, or null if this spliterator cannot be split

Reply:
Can we add a constant for naming this nuul value?